notebooks/notebook_template_review.py

```diff
@@ -193,148 +198,141 @@ def parse_notebook(path: str) -> None:
-    workbench_link = None
-    desc = ''
-    uses = ''
-    steps = ''
-    costs = []
-    
-    @staticmethod
-    def init():
-        NotebookRule.cell_index = 0
-        NotebookRule.title = ''
-        NotebookRule.git_link = None
-        NotebookRule.colab_link = None
-        NotebookRule.workbench_link = None
-        NotebookRule.desc = ''
-        NotebookRule.uses = ''
-        NotebookRule.steps = ''
-        NotebookRule.costs = ''
-    
+    path = None     # The relative path to the notebook
+
     @abstractmethod
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, cells: list) -> list:
         pass
+        
+    def report_error(self,
+                     code: ErrorCode,
+                     errmsg: str) -> None:
+        """
+        Report an error.
+            If args.errors_codes set, then only report these errors. Otherwise, all errors.
+
+        code: The error code number.
+        errmsg: The error message
+        """
+
+        global num_errors
+
+        if args.errors:
+            code = code.value[0]
+            if args.errors_codes:
+                if str(code) not in args.errors_codes:
+                    return
+
+            if args.errors_csv:
+                print(NotebookRule.path, ',', code)
+            else:
+                print(f"{NotebookRule.path}: ERROR ({code}): {errmsg}", file=sys.stderr)
+                num_errors += 1
 
 
 class CopyrightRule(NotebookRule):
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, cells: list) -> list:
         """
         Parse the copyright cell
-
-        path: used only for reporting an error
-        cells: The content cells (JSON) for the notebook
         """
-        cell, cell_index = get_cell(path, cells, NotebookRule.cell_index)
+        cell = cells[0]
         if not 'Copyright' in cell['source'][0]:
-            report_error(path, ErrorCode.ERROR_COPYRIGHT, "missing copyright cell")
-            
-        NotebookRule.cell_index = cell_index
+            self.report_error(ErrorCode.ERROR_COPYRIGHT, "missing copyright cell")
+        cells.pop(0)
+        return cells
 
 
 class NoticesRule(NotebookRule):
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, cells: list) -> list:
         """
         Parse the (optional) notices cell
-
-        path: used only for reporting an error
-        cells: The content cells (JSON) for the notebook
         """
-        cell, cell_index = get_cell(path, cells, NotebookRule.cell_index)
+        cell = cells[0]
         if cell['source'][0].startswith('This notebook'):
-             NotebookRule.cell_index = cell_index
-        else:
-            NotebookRule.cell_index = cell_index - 1
+            cells.pop(0)
+        return cells
 
 
-class TitleRule(NotebookRule):
-    def validate(self, path: str, cells: list) -> None: 
+class TitleRule(NotebookRule): 
+    def validate(self, cells: list) -> (list, str): 
```

Comment:
Method signature has to match the abstractclassethod. Considering using mypy to help you find these issues.

Reply:
was updated